service/security/shiro/src/main/java/org/eclipse/kapua/service/authentication/shiro/realm/UserPassCredentialsMatcher.java

```diff
@@ -179,3 +199,42 @@ public boolean doCredentialsMatch(AuthenticationToken authenticationToken, Authe
     }
 
 }
+
+interface PasswordChecker {
+
+    boolean checkPassword(String tokenPassword, Credential infoCredential);
+
+}
+
+class CachePasswordChecker implements PasswordChecker {
+
+    private static final Cache<String, String> CACHED_CREDENTIALS = new LocalCache<String, String>(
+            KapuaAuthenticationSetting.getInstance().getInt(KapuaAuthenticationSettingKeys.AUTHENTICATION_CREDENTIAL_USERPASS_CACHE_CACHE_SIZE, 1000),
+            KapuaAuthenticationSetting.getInstance().getInt(KapuaAuthenticationSettingKeys.AUTHENTICATION_CREDENTIAL_USERPASS_CACHE_CACHE_TTL, 60),
+            null);
+
+    public boolean checkPassword(String tokenPassword, Credential infoCredential) {
+        String hashedTokenPassword = Base64.getEncoder().encodeToString(DigestUtils.sha3_512(tokenPassword));
+        String cachedCredential = CACHED_CREDENTIALS.get(hashedTokenPassword);
+        if (cachedCredential!=null && cachedCredential.equals(infoCredential.getCredentialKey())) {
+            return true;
+        }
+        else if (BCrypt.checkpw(tokenPassword, infoCredential.getCredentialKey())) {
+            CACHED_CREDENTIALS.put(hashedTokenPassword, infoCredential.getCredentialKey());
```

Comment:
Wouldn't it be more efficient to store them the other way around? retrieve by credential//userId and compare to the password hash? Caching by password hash is more likely to create collisions than the other way around